bookapp/src/error_injection_middleware.rs

```diff
@@ -0,0 +1,397 @@
+pub async fn delete_config(
+    Extension(store): Extension<Arc<dyn ErrorInjectionConfigStore>>,
+    Path(id): Path<i32>,
+) -> Result<StatusCode, StatusCode> {
+    store
+        .delete_config(id)
+        .await
+        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
+    Ok(StatusCode::NO_CONTENT)
+}
+
+/// Middleware that injects errors into requests based on the error injection configurations.
+///
+/// This middleware intercepts incoming requests, checks if there is a matching error injection configuration,
+/// and, based on the error rate, may inject an error response.
+///
+/// # Example Usage
+///
+/// ```rust
+/// use std::sync::Arc;
+/// use axum::{Router, Extension};
+/// use sqlx::PgPool;
+/// use middleware::{error_injection_middleware, PostgresErrorInjectionConfigStore};
+///
+/// fn router(connection_pool: PgPool) -> Router {
+///     // Create the ErrorInjectionConfigStore
+///     let error_injection_store: Arc<dyn ErrorInjectionConfigStore> = Arc::new(
+///         PostgresErrorInjectionConfigStore::new(connection_pool.clone())
+///     );
+///
+///     Router::new()
+///         .layer(Extension(connection_pool))
+///         .layer(Extension(error_injection_store))
+///         .layer(axum::middleware::from_fn(error_injection_middleware))
+///         // ... other routes and layers ...
+/// }
+/// ```
+///
+/// # Arguments
+///
+/// * `req` - The incoming request.
+/// * `next` - The next middleware or handler in the chain.
+///
+/// # Returns
+///
+/// Either an error response or the result of the next middleware/handler.
+#[tracing::instrument(skip_all,
+    fields(
+        method = req.method().to_string(),
+        path = req.uri().path().to_string(),
+        error_rate,
+    )
+)]
+pub async fn error_injection_middleware(
+    State(store): State<Arc<dyn ErrorInjectionConfigStore>>,
+    req: Request,
+    next: Next,
+) -> impl IntoResponse {
+    let path = req.uri().path().to_string();
+    let method = req.method().as_str().to_string();
+
+    // Query the store for matching error injection configurations
+    if let Some(config) = get_matching_error_injection_config(store, &path, &method).await {
+        tracing::Span::current().record("error_rate", &config.error_rate);
+
+        // Generate a random number between 0.0 and 1.0
+        let mut rng = rand::thread_rng();
+        let random_value: f64 = rng.gen();
+
+        if random_value < config.error_rate {
+            tracing::debug!(
+                path = path,
+                method = method,
+                injected_status_code = config.error_code,
+                "Injecting an error"
+            );
+            let status_code = StatusCode::from_u16(config.error_code as u16)
+                .unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
+            let body = config
+                .error_message
+                .unwrap_or_else(|| status_code.canonical_reason().unwrap_or("Error").to_string());
```

Comment:
Would consider a better error message than just `Error`. Maybe `Injected error` to be clear.